bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -49,110 +50,137 @@ class BusinessPartnerService(
-                initSharingState(dto)
-            }
+    fun getBusinessPartnersOutput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerOutputDto> {
+        val stage = StageType.Output
+        return getBusinessPartners(pageRequest, externalIds, stage)
+            .toPageDto(businessPartnerMappings::toBusinessPartnerOutputDto)
     }
 
-    private fun updateBusinessPartnerInput(existingEntity: BusinessPartner, dto: BusinessPartnerInputRequest): BusinessPartner {
-        businessPartnerMappings.updateBusinessPartner(existingEntity, dto, null, null)
-        return businessPartnerRepository.save(existingEntity)
-            .also {
-                saveChangelog(dto.externalId, ChangelogType.UPDATE, StageType.Input, checkBusinessPartnerType(dto.postalAddress.addressType))
-            }
+    private fun upsertBusinessPartnersInput(entityCandidates: List<BusinessPartner>): List<BusinessPartner> {
+        val upsertPartition = replaceWithExistingAndPartition(entityCandidates, StageType.Input)
+
+        saveChangelog(upsertPartition)
+        upsertPartition.allEntities.forEach { entity -> initSharingState(entity) }
+
+        return businessPartnerRepository.saveAll(upsertPartition.allEntities)
+    }
+
+    private fun upsertBusinessPartnersOutput(entityCandidates: List<BusinessPartner>): List<BusinessPartner> {
+        val externalIds = entityCandidates.map { it.externalId }
+        assertInputStageExists(externalIds)
+
+        val upsertPartition = replaceWithExistingAndPartition(entityCandidates, StageType.Output)
+        saveChangelog(upsertPartition)
+
+        return businessPartnerRepository.saveAll(upsertPartition.allEntities)
+    }
+
+    private fun getBusinessPartners(pageRequest: PageRequest, externalIds: Collection<String>?, stage: StageType): Page<BusinessPartner> {
+        return when {
+            externalIds.isNullOrEmpty() -> businessPartnerRepository.findByStage(stage, pageRequest)
+            else -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
+        }
+    }
+
+
+    private fun copyValues(fromPartner: BusinessPartner, toPartner: BusinessPartner): BusinessPartner {
+        toPartner.stage = fromPartner.stage
+        toPartner.nameParts = fromPartner.nameParts
+        toPartner.roles = fromPartner.roles
+        toPartner.identifiers = fromPartner.identifiers
+        toPartner.states = fromPartner.states
+        toPartner.classifications = fromPartner.classifications
+        toPartner.shortName = fromPartner.shortName
+        toPartner.legalForm = fromPartner.legalForm
+        toPartner.isOwner = fromPartner.isOwner
+        toPartner.bpnL = fromPartner.bpnL
+        toPartner.bpnS = fromPartner.bpnS
+        toPartner.bpnA = fromPartner.bpnA
+        toPartner.postalAddress = fromPartner.postalAddress
```

Comment:
I'm not sure if we want to replace the whole `PostalAddress` with a new one. At least orphans are removed via annotation.
This works, but it's different to the way we handle `BusinessPartner`s. Theoretically we could also delete them and create new ones on update.
Alternative would be a `copyValues` for `PostalAddress`.
What do you think is preferable?

Reply:
I decided to add copyValues methods for all independent entities attached to business partner entity